### DISCUSSION TITLE
Allow caching artifacts without privileged containers and volumes

### DISCUSSION THREAD

**Discussion by 9p4** [woodpecker-ci/woodpecker#2296] [Request a Feature]
### Clear and concise description of the problem

Allow for the pipeline to cache artifacts, much like what Travis CI allows you to do: https://docs.travis-ci.com/user/caching/

This is currently possible in Woodpecker, but only if using volumes. It would be pretty neat if volumes weren't required and Woodpecker could cache items without exposing security issues.

### Suggested solution

When the pipeline completes, save the indicated directories somewhere. Then, on subsequent runs of the pipeline, mount that directory automatically. Perhaps automatic volume creation?

### Alternative

The current method is insecure and should not be available to untrusted users.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by anbraten** [comment]
Caching is a highly requested feature and we talked a lot about it already. I would prefer to have some official plugins which help the user to do caching. How the caching works really depends on the "backend" you want to use IMO. Most users will probably be fine with simply putting caches in a folder on the server, but as you can have multiple agents for your Woodpecker instance or could even use Kubernetes (#9) as backend, most larger deployments probably need some kind of clustered caching backend like s3 as well. Woodpecker should probably be the generic base for it and the caching plugins could do the magic in the background. If we have some nice plugins we should add a sample about them somewhere in the docs as well (#316).

**Comment by wxiaoguang** [comment]
Although putting cache into S3 is a choice but it is still slow and wastes bandwidth.

In some cases, we have predictable hosts to run agents. So on every host, we can cache the files separately. I do not care about if the cache is sync-ed. If there is cache, then I'd like to reuse it; if there is no cache, then I can let the job to init a new cache, instead of restoring it from somewhere else.


ps: now I am using such local cache mechanism, and I use Docker-in-Docker to isolate the agents from hosts, then I can set `trusted` to all repositories. But if the CI agent could support local cache, then users do not need to set `trusted` anymore, and do not need to setup Docker-in-Docker.

**Comment by 6543** [comment]
"Allow agent to cache" would be the feature then I guess?

**Reply by pinpox** [reply]
This would still be relevant for me. I'm runing a build step which requires a big download as prerequisite. It would speed up my action a lot and save bandwidth if I could cache the downloaded files and only download them again if not present.

**Reply by ficd0** [reply]
Agreed. 2026 and this is still of interest.
